docs/proposals/202005_scalable-rule-storage.md

```diff
@@ -0,0 +1,68 @@
+---
+title: Scalable rule storage
+type: proposal
+menu: proposals
+status: proposed
+---
+
+## Summary
+
+There is no way to scale rule evaluation and storage today except functionally
+sharding rules onto multiple instances the `thanos rule` component. However, we
+have already solved scaling storage of time-series across multiple processes:
+`thanos receive`.
+
+To scale storing results of rule evaluations this proposal proposes to allow
+the `thanos rule` component to have a stateless mode, storing results of
+queries by sending it to a `thanos receive` hashring instead of storing them
+locally.
+
+## Motivation
+
+A few large rules can create a significant amount of resulting time-series,
+which should not be limited to what we can scale to with a single TSDB embedded
+in the rule component.
+
+## Goals
+
+Allow scaling storage of rule evaluations.
+
+## Verification
+
+* Run all rule component e2e tests with new mode as well.
+
+## Proposal
+
+Allow specifying one of the following flags:
+
+* `--remote-storage`
+* `--remote-storage.config` or `--remote-storage.config-file` flag following the same scheme as
+[`--query.config`, and
+`--query.config-file`](https://thanos.io/components/rule.md/#query-api)
+
+If any of these are specified the ruler would run a stateless mode, without
+local storage, and instead writing samples to the configured remote server,
+which must implement the `storepb.WritableStore` gRPC service.
+
+## Alternatives
+
+Continue to allow spreading load only by functionally sharding rules.
+
+## Work Plan
+
+Implement functionality alongside the existing architecture of the rule
+component, and once tried and tested we can potentially remove today's
+architecture.
+
```

Comment:
what will we remove here? I think embedded TSDB is here to stay?

Reply:
I'll move this to open questions instead. I think for now we are not touching the embedded TSDB approach, this may be reevaluated in the future though.